### ISSUE TITLE
Add optional embedded local LLM with hardware detection and in-process inference

### ISSUE THREAD

**Issue by acoliver** [acoliver/personal-agent#130]
## Summary

Add an **optional embedded local LLM** that runs in-process, giving users immediate, friction-free AI access without API keys or network calls. This is additive — cloud APIs remain fully supported. The local model is a "zero-config" option that "just works" for users with capable hardware.

## Goals

1. **Hardware detection** — automatically detect if the user’s machine can run a small local model (RAM, GPU/CPU)
2. **Friction-free setup** — offer to download a capable default model (e.g., Qwen3.5 4B) with one click
3. **In-process runtime** — model runs inside PersonalAgent via Candle (pure Rust), no external server
4. **Seamless profile integration** — local model appears as just another provider alongside OpenAI, Anthropic, etc.
5. **Zero friction for cloud users** — existing API workflows unchanged; this is additive, not replacement

## Why This Matters

- **Immediate gratification**: New users can chat immediately without signing up for APIs
- **Privacy**: Conversations stay local
- **Offline capability**: Works without internet
- **Cost**: Zero ongoing API costs for local usage
- **Latency**: No network round-trips for inference

### 1. Hardware Capability Detection

On startup (and in settings), detect whether the system can run a local model:

```rust
pub struct HardwareCapabilities {
    pub total_ram_gb: f64,
    pub available_ram_gb: f64,
    pub has_gpu: bool,
    pub gpu_vram_gb: Option<f64>,
    pub cpu_cores: usize,
    pub is_apple_silicon: bool,  // Metal acceleration available
}

pub fn can_run_local_model(hw: &HardwareCapabilities) -> bool {
    // For Qwen3.5 4B Q4_K_M (~3.4GB file, ~4-5GB runtime)
    hw.available_ram_gb >= 6.0 || hw.gpu_vram_gb.map_or(false, |v| v >= 4.0)
}
```

**Detection sources:**
- macOS: `sysctl hw.memsize`, Metal GPU queries
- Linux: `/proc/meminfo`, `nvidia-smi` or ROCm tools
- Windows: WMI queries, CUDA/ROCm detection

**User-visible:** Show capability status in Settings → Local Model with a clear "Your machine can/cannot run local models" message.

### 2. Default Model Download Flow

If hardware is capable and no local model is present:

**First-run offer:**
```
┌─────────────────────────────────────────┐
│  Run AI locally?                        │
│                                         │
│  Your Mac has 16GB RAM — you can run    │
│  a small AI model directly on your      │
│  machine for offline, private use.      │
│                                         │
│  [Download Qwen3.5 4B (3.4 GB)]         │
│  [Not now]                              │
│                                         │
│  You can always use cloud APIs instead  │
│  or download this later in Settings.    │
└─────────────────────────────────────────┘
```

**Download specs:**
| Attribute | Value |
|-----------|-------|
| Default model | Qwen3.5 4B Instruct |
| Quantization | Q4_K_M |
| File size | ~3.4 GB |
| Download source | HuggingFace (configurable mirror) |
| Resume support | Yes (HTTP range requests) |
| Verify checksum | SHA256 |

**Storage location:**
- macOS: `~/Library/Application Support/PersonalAgent/models/`
- Linux: `~/.local/share/personalagent/models/`
- Windows: `%APPDATA%/PersonalAgent/models/`

### 3. In-Process Inference (Candle)

Use HuggingFace Candle for pure-Rust inference:

```rust
// src/llm/local/candle_provider.rs
pub struct CandleLocalModel {
    model: quantized_llama::Model,
    tokenizer: Tokenizer,
    device: Device,
    config: LocalModelConfig,
}

impl LlmProvider for CandleLocalModel {
    async fn chat(
        &self,
        messages: &[ChatMessage],
        tools: Option<&[Tool]>,
    ) -> Result<LlmResponse> {
        // Tokenize
        // Forward through model
        // Parse output (including tool calls)
        // Return structured response
    }
}
```

**Key implementation points:**
- **Lazy loading**: Load model on first use, not at startup
- **Lazy download**: Download model on first use if user opted in
- **Chat template**: Apply Qwen3.5 chat template (or auto-detect from GGUF metadata)
- **Tool calling**: Wrap tools in prompt, parse `<function=name>{...}</function>` or JSON from output
- **Context window**: Respect model limits (32K for Qwen3.5 4B)
- **Threading**: Run inference on dedicated thread pool to avoid blocking GPUI

### 4. Profile Integration

Local model appears as a first-class provider:

```rust
// In ModelProfile
pub enum Provider {
    OpenAi { api_key: String, base_url: Option<String> },
    Anthropic { api_key: String },
    Local { model_path: PathBuf },
}
```

**Model Selector UI:**
```
┌──────────────────────────────────────┐
│  Select Model                         │
│                                      │
│  ═══ Local Models ═══                │
│  ⬜ Qwen3.5 4B (Local) ✓ Downloaded  │
│                                     │
│  ═══ Cloud Providers ═══            │
│  ⬜ Claude 3.5 Sonnet (Anthropic)     │
│  ⬜ GPT-4o (OpenAI)                  │
│  ⬜ ...                               │
└──────────────────────────────────────┘
```

**Profile creation:**
- If local model downloaded, offer "Quick Start: Local AI" profile
- User can create additional profiles with different local models
- Cloud API profiles unchanged

### 5. Advanced: Bring Your Own Model

Power users can use other GGUF models:

**Settings → Local Model:**
- Download default model button
- "Add custom model" (select GGUF file or download URL)
- Model management (delete to free space, verify integrity)
- Show disk usage

**Compatible formats:**
- GGUF (preferred)
- Model must fit in detected RAM/VRAM
- Tool calling quality varies by model (show user rating/benchmarks)

### Memory handling

| Scenario | Behavior |
|----------|----------|
| User has 8GB RAM | Offer model but warn about memory pressure |
| User has 16GB+ RAM | Default offer enabled |
| Model running, RAM pressure | Allow manual unload, auto-unload on low memory (optional) |
| App quit | Model unloads automatically |

### GPU acceleration

- **Apple Silicon**: Use Metal via `candle-metal` or `metal-candle`
- **NVIDIA**: Use CUDA if available
- **CPU**: Works but slower (acceptable for lightweight use)

### Scenario A: Capable machine, first run

1. App launches
2. Hardware detection runs briefly in background
3. "Your machine can run AI locally" notification/banner appears
4. User clicks "Set up local AI"
5. One-click download with progress bar
6. "Local AI ready" notification
7. New "Local" profile available in model selector
8. User can chat immediately without any API keys

### Scenario B: Cloud-first user

1. User ignores local offer or clicks "Not now"
2. App functions exactly as today
3. Later, user can go to Settings → Local Model to download
4. No pressure, no degraded experience

### Scenario C: Low-resource machine

1. Hardware detection determines 4GB RAM
2. Settings shows "Your machine doesn’t meet requirements for local AI (needs 6GB+ RAM)"
3. Offer link to learn more or use cloud providers
4. No download button shown

### Crate dependencies

```toml
[dependencies]
# Optional local model feature
candle-core = { version = "0.9", optional = true }
candle-transformers = { version = "0.9", optional = true }
tokenizers = { version = "0.21", optional = true }

[features]
default = []
local-model = ["candle-core", "candle-transformers", "tokenizers"]
```

### Model format

- **Input**: GGUF files (industry standard for quantized local models)
- **Quantization**: Q4_K_M for default (good quality, small size)
- **Larger variants**: Support Q5_K_M or Q8_0 if user wants quality over size

### Chat template

Parse GGUF metadata for `tokenizer.chat_template` field and apply via Candle’s templating.

### Tool calling prompt format

```
You are a helpful assistant with access to tools.

Available tools:
- read_file(path: string): Read file contents
- search(query: string): Search the web

Respond with chat, or use a tool like:
<function=read_file>{"path": "README.md"}</function>
```

Parse `<function=name>{...}</function>` or JSON from model output.

## Acceptance Criteria

- [ ] Hardware capability detection (RAM, GPU) implemented for macOS/Linux/Windows
- [ ] Detection runs at startup and shows status in Settings
- [ ] Default model download flow implemented (Qwen3.5 4B or equivalent)
- [ ] Download resumes on interruption, verifies checksum
- [ ] Candle-based in-process inference working
- [ ] Local model appears in model selector alongside cloud providers
- [ ] User can chat with local model without any API configuration
- [ ] Settings allows downloading default model later if skipped initially
- [ ] Settings allows adding custom GGUF models
- [ ] Settings shows model management (disk usage, delete, verify)
- [ ] Lazy loading: model only loaded into memory when first used
- [ ] Model unloads cleanly on app quit or manual unload
- [ ] Cloud API providers continue to work unchanged
- [ ] Low-resource machines show friendly "not supported" message instead of offer

## Out of Scope (Future Work)

- Model fine-tuning or training
- Automatic model selection based on task
- Multiple local models loaded simultaneously
- Model quantization/conversion in-app
- Distributed/combined local + cloud inference

## Related Research

- Qwen3.5 4B tool-calling eval: 97.5% accuracy (March 2026)
- Candle ML framework: pure Rust, 3x faster than PyTorch
- GGUF format: industry standard for quantized local models
- Hardware detection: standard system APIs on all platforms

| Model | Size | Tool Calling | Speed | Best For |
|-------|------|--------------|-------|----------|
| Qwen3.5 4B | 3.4GB | 97.5% | 48 tok/s | Default embedded |
| Nemotron Nano 4B | 4.2GB | 95% | 44 tok/s | Sequential tasks |
| FunctionGemma 270M | ~0.5GB | ~75% | 60+ tok/s | Ultra-low resource |

## Notes

- This feature is **additive** — it does not replace or compete with cloud providers
- Goal is "works out of the box for capable hardware" not "maximum performance"
- Prioritize reliability and user experience over raw speed
- One default model is enough; power users can bring their own

**Comment by acoliver** [comment]
@coderabbit looks like what we needed was this: https://github.com/woodRock/candle/tree/feat-qwen3.5-support -

**Comment by m13v** [comment]
hit the exact qwen gguf token salad with candle last summer in a mac app. candle's quantized_llama path broke on the newer rope scaling configs, and the woodrock fork works but lags mainline by weeks whenever qwen pushes a new variant. what actually worked for shipping: llama-cpp-rs via FFI on intel macs, mlx-rs on apple silicon. mlx was 2-3x faster than candle on an m2 pro with qwen 4b q4. also don't gate on total ram alone, unified memory on 8gb m-series chokes even a 4b model once safari is open. check available_ram and wired_count from vm_stat.
